core/src/service/api-creator.ts

```diff
@@ -0,0 +1,230 @@
+  autoMessage?: boolean; // eject message automatically
+  successMsg?: string; // eject message when success to override default message
+  errorMsg?: string; // eject message when failed to override default message
+}
+type $headers = Record<string, string>;
+type $body = Record<string, any>;
+interface CallParams {
+  $options?: $options;
+  $headers?: $headers;
+  $body?: $body;
+}
+
+type Merge<A, B> = { [K in keyof A]: K extends keyof B ? B[K] : A[K] } & B extends infer O
+  ? { [K in keyof O]: O[K] }
+  : never;
+
+/**
+ * generate api request function by config
+ * @param apiConfig
+ * @returns callable api function
+ */
+export const genRequest = function <T extends FN>(apiConfig: APIConfig) {
+  const { api, headers } = apiConfig;
+  let [method, path] = api.split('@');
+  if (!path) {
+    path = method;
+    method = 'get';
+  }
+
+  return (params: CallParams & Merge<Parameters<T>[0], {}>) => {
+    const { $options, $headers, $body, ...rest } = params;
+    const { bodyOrQuery, pathParams } = extractPathParams(path, rest);
+    const { isDownload, uploadFileKey } = $options || {};
+    if ('pageNo' in bodyOrQuery && !('pageSize' in bodyOrQuery)) {
+      bodyOrQuery.pageSize = DEFAULT_PAGESIZE;
+    }
+    let bodyData;
+    if (['post', 'put'].includes(method)) {
+      bodyData = isEmpty(bodyOrQuery) ? undefined : uploadFileKey ? bodyOrQuery[uploadFileKey] : bodyOrQuery;
+    } else if (method === 'delete') {
+      bodyData = $body;
+    }
+    return axios({
+      method: method as any,
+      url: generatePath(path, pathParams),
+      headers: !isEmpty(headers) ? headers : $headers,
+      params: bodyOrQuery,
+      paramsSerializer: (p: Record<string, string>) => qs.stringify(p),
+      responseType: isDownload ? 'blob' : 'json',
+      data: bodyData,
+    }).then((res) => res.data) as unknown as Promise<BODY<ReturnType<T>>>;
+  };
+};
+
+export const apiDataStore = createStore({
+  name: 'api_data',
+  state: {
+    body: {} as Obj,
+    data: {} as Obj,
+    loading: {} as Obj<boolean>,
+  },
+  reducers: {
+    setBody(state, path: string, data: any) {
+      state.body[path] = data;
+    },
+    setData(state, path: string, data: any) {
+      state.data[path] = data;
+    },
+    setLoading(state, path: string, isLoading: boolean) {
+      state.loading[path] = isLoading;
+    },
+  },
+});
+
+type FN = (...args: any) => any;
+type RES<D> = Promise<BODY<D>>;
+type PICK_BODY<T extends FN> = ReturnType<T> extends RES<infer D> ? BODY<D> : never;
+type PICK_DATA<T extends FN> = ReturnType<T> extends RES<infer D> ? D : never;
+interface APIConfig {
+  api: string;
+  successMsg?: string;
+  errorMsg?: string;
+  paging?: any;
+  globalKey?: string;
+  headers?: Record<string, any>;
+}
+interface BODY<D> {
+  data: D | null;
+  success: boolean;
+  err: {
+    msg: string;
+    code: string;
+    ctx: null | object;
+  };
+  userInfo?: Record<string, object>;
+}
+
+const noop = (d: any) => {};
```

Comment:
optional call  instead of noop

Reply:
It's used before replaced with real hook setState.